File: 프로그래머스/1/172928. 공원 산책/공원 산책.js
const DIRECTION_MAP = {
    E: [0, 1],
    W: [0, -1],
    S: [1, 0],
    N: [-1, 0],
  };


const solution = (park, routes) => {
  park = park.map((p) => p.split(''));

  const n = park.length;
  const m = park[0].length;

  let pos = getStartPos();
    
  function getStartPos() {
    for (let i = 0; i < n; i++) {
      for (let j = 0; j < m; j++) {
        if (park[i][j] === 'S') {
          return new Position(i, j);
        }
      }
    }
  }

  function canMove(pos, dir, step) {
    const tempPos = new Position(pos.x, pos.y);

    for (let i = 0; i < step; i++) {
      tempPos.move(DIRECTION_MAP[dir], 1);
      if (tempPos.isLimit(n, m) || tempPos.isBlocked(park)) return false;
    }
    return true;
  }

  for (const route of routes) {
    const routeInfo = route.split(' ');
    const [dir, step] = routeInfo;
    if (canMove(pos, dir, step)) {
      pos.move(DIRECTION_MAP[dir], step);
    }
  }

  return [pos.x, pos.y];
};

class Position {
  constructor(x, y) {
    this.x = x;
    this.y = y;
  }

  move(direction, step) {
    this.x += direction[0] * step;
    this.y += direction[1] * step;
  }

  isLimit(r, c) {
    return this.x < 0 || this.x >= r || this.y < 0 || this.y >= c;
  }

  isBlocked(board) {
    return board[this.x][this.y] === 'X';
  }
}

File: 프로그래머스/1/172928. 공원 산책/공원 산책.java
import java.util.HashMap;
import java.util.Map;


class Solution {
    public int[] solution(String[] park, String[] routes) {
        Map<String, int[]> direction = new HashMap<>();
        
        direction.put("E", new int[]{0, 1});
        direction.put("W", new int[]{0, -1});
        direction.put("S", new int[]{1, 0});
        direction.put("N", new int[]{-1, 0});
        
        int x = 0; 
        int y = 0;
        
        for(int i = 0; i < park.length; i++){
            for(int j = 0; j < park[0].length(); j++){
                if(park[i].charAt(j) == 'S'){
                    x = i;
                    y = j;
                }
            }
        }
        
        
  
        
        for(String route : routes){
            String[] splited = route.split(" ");
            String r = splited[0];
            int n = Integer.parseInt(splited[1]);
            
            int nx = x;
            int ny = y;
            
            int step = 0;
            
            while(step < n) {
                nx = nx + direction.get(r)[0];
                ny = ny + direction.get(r)[1];
                
                if(nx < 0 || nx >= park.length || ny < 0 || ny >= park[0].length() || park[nx].charAt(ny) == 'X') break;
                step++;        
            }
            
            if(step == n) {
                x = nx;
                y = ny;
            }
        }
    
      return new int[]{x, y};
    }
}